// playwright/lib/mcp/sdk/mdb.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mdb_exports = {};
__export(mdb_exports, {
  MDBBackend: () => MDBBackend,
  runMainBackend: () => runMainBackend,
  runOnPauseBackendLoop: () => runOnPauseBackendLoop
});
module.exports = __toCommonJS(mdb_exports);
var import_utilsBundle = require("playwright-core/lib/utilsBundle");
var import_utils = require("playwright-core/lib/utils");
var import_tool = require("./tool");
var mcpBundle = __toESM(require("./bundle"));
var mcpServer = __toESM(require("./server"));
var mcpHttp = __toESM(require("./http"));
var import_server = require("./server");
const mdbDebug = (0, import_utilsBundle.debug)("pw:mcp:mdb");
const errorsDebug = (0, import_utilsBundle.debug)("pw:mcp:errors");
const z = mcpBundle.z;
class MDBBackend {
  constructor(topLevelBackend, allowedOnPause) {
    this._stack = [];
    this._progress = [];
    this._topLevelBackend = topLevelBackend;
    this._allowedOnPause = allowedOnPause;
  }
  async initialize(server, clientInfo) {
    if (!this._clientInfo)
      this._clientInfo = clientInfo;
  }
  async listTools() {
    const client = await this._client();
    const response = await client.listTools();
    return response.tools;
  }
  async callTool(name, args) {
    await this._client();
    if (name === pushToolsSchema.name)
      return await this._pushTools(pushToolsSchema.inputSchema.parse(args || {}));
    const interruptPromise = new import_utils.ManualPromise();
    this._interruptPromise = interruptPromise;
    if (!this._allowedOnPause.includes(name)) {
      for (let i = 0; i < this._stack.length - 1; i++) {
        if (this._stack[i].toolNames.includes(name))
          break;
        await this._stack[i].client.close().catch(errorsDebug);
        this._stack.shift();
      }
    }
    let entry;
    for (let i = 0; i < this._stack.length; i++) {
      if (this._stack[i].toolNames.includes(name)) {
        entry = this._stack[i];
        break;
      }
      mdbDebug("popping client from stack for ", name);
    }
    if (!entry)
      throw new Error(`Tool ${name} not found in the tool stack`);
    const client = entry.client;
    const resultPromise = new import_utils.ManualPromise();
    entry.resultPromise = resultPromise;
    client.callTool({
      name,
      arguments: args,
      _meta: {
        progressToken: name + "@" + (0, import_utils.createGuid)().slice(0, 8)
      }
    }).then((result2) => {
      resultPromise.resolve(result2);
    }).catch((e) => {
      resultPromise.resolve({ content: [{ type: "text", text: String(e) }], isError: true });
    });
    const result = await Promise.race([interruptPromise, resultPromise]);
    if (interruptPromise.isDone())
      mdbDebug("client call intercepted", result);
    else
      mdbDebug("client call result", result);
    result.content.unshift(...this._progress);
    this._progress.length = 0;
    return result;
  }
  async _client() {
    if (!this._stack.length) {
      const transport = await (0, import_server.wrapInProcess)(this._topLevelBackend);
      await this._pushClient(transport);
    }
    return this._stack[0].client;
  }
  async _pushTools(params) {
    mdbDebug("pushing tools to the stack", params.mcpUrl);
    const transport = new mcpBundle.StreamableHTTPClientTransport(new URL(params.mcpUrl));
    await this._pushClient(transport, params.introMessage);
    return { content: [{ type: "text", text: "Tools pushed" }] };
  }
  async _pushClient(transport, introMessage) {
    mdbDebug("pushing client to the stack");
    const client = new mcpBundle.Client({ name: "Interrupting client", version: "0.0.0" }, { capabilities: { roots: {} } });
    client.setRequestHandler(mcpBundle.ListRootsRequestSchema, () => ({ roots: this._clientInfo?.roots || [] }));
    client.setRequestHandler(mcpBundle.PingRequestSchema, () => ({}));
    client.setNotificationHandler(mcpBundle.ProgressNotificationSchema, (notification) => {
      if (notification.method === "notifications/progress") {
        const { message } = notification.params;
        if (message)
          this._progress.push({ type: "text", text: message });
      }
    });
    await client.connect(transport);
    mdbDebug("connected to the new client");
    const { tools } = await client.listTools();
    this._stack.unshift({ client, toolNames: tools.map((tool) => tool.name), resultPromise: void 0 });
    mdbDebug("new tools added to the stack:", tools.map((tool) => tool.name));
    mdbDebug("interrupting current call:", !!this._interruptPromise);
    this._interruptPromise?.resolve({
      content: [{
        type: "text",
        text: introMessage || ""
      }]
    });
    this._interruptPromise = void 0;
    return { content: [{ type: "text", text: "Tools pushed" }] };
  }
}
const pushToolsSchema = (0, import_tool.defineToolSchema)({
  name: "mdb_push_tools",
  title: "Push MCP tools to the tools stack",
  description: "Push MCP tools to the tools stack",
  inputSchema: z.object({
    mcpUrl: z.string(),
    introMessage: z.string().optional()
  }),
  type: "readOnly"
});
async function runMainBackend(backendFactory, allowedOnPause, options) {
  const mdbBackend = new MDBBackend(backendFactory.create(), allowedOnPause);
  const factory = {
    ...backendFactory,
    create: () => mdbBackend
  };
  const url = await startAsHttp(factory, { port: options?.port || 0 });
  process.env.PLAYWRIGHT_DEBUGGER_MCP = url;
  if (options?.port !== void 0)
    return url;
  await mcpServer.connect(factory, new mcpBundle.StdioServerTransport(), false);
}
async function runOnPauseBackendLoop(backend, introMessage) {
  const wrappedBackend = new ServerBackendWithCloseListener(backend);
  const factory = {
    name: "on-pause-backend",
    nameInConfig: "on-pause-backend",
    version: "0.0.0",
    create: () => wrappedBackend
  };
  const httpServer = await mcpHttp.startHttpServer({ port: 0 });
  await mcpHttp.installHttpTransport(httpServer, factory);
  const url = mcpHttp.httpAddressToString(httpServer.address());
  const client = new mcpBundle.Client({ name: "Pushing client", version: "0.0.0" });
  client.setRequestHandler(mcpBundle.PingRequestSchema, () => ({}));
  const transport = new mcpBundle.StreamableHTTPClientTransport(new URL(process.env.PLAYWRIGHT_DEBUGGER_MCP));
  await client.connect(transport);
  const pushToolsResult = await client.callTool({
    name: pushToolsSchema.name,
    arguments: {
      mcpUrl: url,
      introMessage
    }
  });
  if (pushToolsResult.isError)
    errorsDebug("Failed to push tools", pushToolsResult.content);
  await transport.terminateSession();
  await client.close();
  await wrappedBackend.waitForClosed();
  httpServer.close();
}
async function startAsHttp(backendFactory, options) {
  const httpServer = await mcpHttp.startHttpServer(options);
  await mcpHttp.installHttpTransport(httpServer, backendFactory);
  return mcpHttp.httpAddressToString(httpServer.address());
}
class ServerBackendWithCloseListener {
  constructor(backend) {
    this._serverClosedPromise = new import_utils.ManualPromise();
    this._backend = backend;
  }
  async initialize(server, clientInfo) {
    await this._backend.initialize?.(server, clientInfo);
  }
  async listTools() {
    return this._backend.listTools();
  }
  async callTool(name, args, progress) {
    return this._backend.callTool(name, args, progress);
  }
  serverClosed(server) {
    this._backend.serverClosed?.(server);
    this._serverClosedPromise.resolve();
  }
  async waitForClosed() {
    await this._serverClosedPromise;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MDBBackend,
  runMainBackend,
  runOnPauseBackendLoop
});
